Add search index data

p-galligan opened the pull request:
Adds search index data.

Currently a little stuck on how to deal with notes in a systematic manner. There are a huge variety, and I'm not sure we'd need all of them. I figure Publication and ISBN maybe? Additionally, I'm not quite sure how to capture them with the proper title/content as for the others, I'm basically just writing strings to an array in liquid.

Additionally, a little stumped on how to work with top containers. We really only want these for the call number, which is in the indicator. Some library resources have multiple top containers, but not all of them have the call number (I think), and some are multiple top containers with the same call number... See https://dimes.rockarch.org/xtf/view?docId=mods/LI03963/LI03963.xml;query=;chunk.id=headerlink;brand=default and https://dimes.rockarch.org/xtf/view?docId=mods/LI07343/LI07343.xml;query=;chunk.id=headerlink;brand=default as examples.

Some guidance on whether what I have is the correct approach for this, and how to handle top containers/notes would be appreciated

Commit:
skeleton search data

add some search data for indexing

helrond commented on search_data.json:
```diff
@@ -0,0 +1,50 @@
+    {% assign object = item_hash[1] %}
+    {% assign id = item_hash[0] %}
+    {% assign collection_id = object.uri | split:'/' | last %}
+    {% assign subjects = "" | split: ',' %}
+    {% assign agents = "" | split: ',' %}
+    {% assign dates = "" | split: ',' %}
```
You're creating empty list variables here, then pushing into them below, right? Just want to confirm my understanding of what's going on here.

p-galligan replied:
Yeah, that's what's happening. I figured we just needed a list of the names for subjects and agents without the other info, but I could be wrong.

helrond commented on search_data.json:
```diff
@@ -0,0 +1,50 @@
+    {% assign agents = "" | split: ',' %}
+    {% assign dates = "" | split: ',' %}
+
+    {% for date in object.dates %}
+      {% assign dates = dates | push: date.expression %}
+    {% endfor %}
```
In lieu of a date facet or ability to filter by date, I wonder about the utility of indexing date data?

p-galligan replied:
That's a good point. I was including the dates because they are technically the publication dates. But I'm happy to remove them since we don't have a date filter.

helrond reviewed (commented):
I think you're definitely headed in the right direction! I've added a couple of specific comments just to make sure I understand what's going on.

More global comments:
- It might be good to look at stripping whitespace out of this file so it's as compact as possible. See https://shopify.github.io/liquid/basics/whitespace/
- For note parsing, you can look at https://github.com/RockefellerArchiveCenter/VirtualVault/blob/base/site/_includes/note.html. It won't get everything, but for our purposes it may be good enough.
- For call numbers, it looks like there's a `uniq` filter in Liquid, so you could easily just get an array of everything, filter it with that function, and then join the remainder? I am curious though if the number of instances corresponds to the number of copies of a book that we have? That would be good to know for the detail template.

helrond commented on search_data.json:
```diff
@@ -0,0 +1,50 @@
+layout: null
+---
+{
+  {% for item_hash in data.resource %}
```
also, this needs to `site.data.resource`

helrond commented on search_data.json:
```diff
@@ -0,0 +1,50 @@
+      "title": "{{object.title | escape | strip_newlines | remove: '\'}}",
+      "url": "{{site.url}}/resource/{{id}}",
+      "dates": "{{ dates }}",
+      "agents": "{{ agents }}",
+      "subjects": "{{ subjects }}",
```
You'll need to do something so you have a space in between items in the array.  See https://shopify.github.io/liquid/filters/join/

Commit:
update index

Updates index to remove whitespace, dates, and add a join separator for agents and subjects.

Commit:
add call numbers

Add an array for unique call numbers

Commit:
add note logic

adds logic to get notes

Commit:
add note capture

Adds logic to capture note label and content.

Removes chronology check because we don't have any in library resources.

helrond commented on search_data.json:
```diff
@@ -0,0 +1,65 @@
+        {% endfor %}
+      {% when 'note_singlepart' %}
+        {%- assign content = note.content -%}
+      {% endcase %}
```
Can we move this to an include called note_content or something so we can repurpose it for notes in item details?

helrond commented on search_data.json:
```diff
@@ -0,0 +1,65 @@
+      {%- endcapture -%}
+      {%- assign obj = site.data.top_container[top_container_id] -%}
+      {%- assign call_numbers = call_numbers | push: obj.indicator -%}
+    {%- endfor -%}
```
Ditto here - would be great to have an include we can use in detail pages.

helrond commented on search_data.json:
```diff
@@ -0,0 +1,65 @@
+      "agents": "{{ agents | join: ' | ' }}",
+      "subjects": "{{ subjects | join: ' | ' }}",
+      "notes": "{{ notes | join: ', '}}",
+      "call_numbers": "{{ call_numbers | uniq | join ' | '}}"
```
Curious about the thinking in join characters - is this going to matter in indexing?

p-galligan replied:
The only thinking I made with join characters here was that a simple comma would be confusing specifically for agents, and subjects possibly with name forms. However, they are not related to indexing specifically, and I could got with `, ` instead.

helrond commented on search_data.json:
```diff
@@ -0,0 +1,65 @@
+      "agents": "{{ agents | join: ' | ' }}",
+      "subjects": "{{ subjects | join: ' | ' }}",
+      "notes": "{{ notes | join: ', '}}",
+      "call_numbers": "{{ call_numbers | uniq | join ' | '}}"
```
Missing a colon here

helrond commented on search_data.json:
```diff
@@ -0,0 +1,65 @@
+
+    {%- assign notes = "" | split: ',' -%}
+    {%- for note in object.notes -%}
+      {%- assign label = note.label -%}
```
I wonder if we need to index note labels or if just the content is good enough? If we do index labels, you'll need to add a space after the colon in line 42 below.

helrond commented on search_data.json:
```diff
@@ -0,0 +1,65 @@
+      {% when 'note_multipart' %}
+        {% for subnote in note.subnotes %}
+          {% if subnote.jsonmodel_type == 'note_text' %}
+            {%- assign content = subnote.content -%}
```
This should probably strip out trailing returns, as I'm seeing a lot of those in the generated data.

helrond commented on search_data.json:
```diff
@@ -0,0 +1,65 @@
+    "resource/{{id}}": {
+      "href": "resource/{{id}}",
+      "title": "{{object.title | escape | strip_newlines | remove: '\'}}",
+      "url": "{{site.url}}/resource/{{id}}",
```
`url` and `href` are the same value...

p-galligan replied:
I'm leaning towards the full url instead of just the end href. Does that make sense?

helrond reviewed (changes requested):
OK, I would encourage you to take a closer look at search-data.json after it's built, because there are a number of things that could be improved. I commented on some of those here, but a closer look at that file (and running it across more than just the sample data) would be a good idea.

I've also requested moving some of this logic into includes so we can reuse it on item detail pages.

Commit:
Adds missing top containers

Adds missing top containers.

Makes includes for note content and top containers.

Removes labels from note index.

Changes join character to `, `.

helrond committed:
being annoying with formatting

helrond reviewed (approved):
Right on!
